main/src/library/Nec.flix

```diff
@@ -0,0 +1,909 @@
+    }
+
+    ///
+    /// Returns a list of every element in `c` that satisfies the predicate `f`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def filter(f: a -> Bool & ef, c: Nec[a]): List[a] & ef =
+        def loop(c1: Nec[a], k: List[a] -> List[a] & ef) = match viewLeft(c1) {
+            case OneLeft(x)      => if (f(x)) k(x :: Nil) else k(Nil)
+            case SomeLeft(x, rs) => if (f(x)) loop(rs, ks -> k(x :: ks)) else loop(rs, k)
+        };
+        loop(c, ks -> ks as & ef)
+
+
+
+    ///
+    /// Returns the result of applying `combine` to all the elements in `l`, using `empty` as the initial value.
+    ///
+    pub def fold(l: Nec[a]): a with Monoid[a] = Foldable.fold(l)
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `c` going from left to right.
+    ///
+    /// That is, the result is of the form: `f(...f(f(s, x1), x2)..., xn)`.
+    ///
+    pub def foldLeft(f: (b, a) -> b & ef, acc: b, c: Nec[a]): b & ef = match viewLeft(c) {
+        case OneLeft(x)      => f(acc, x)
+        case SomeLeft(x, rs) => {
+            let b = f(acc, x);
+            foldLeft(f, b, rs)
+        }
+    }
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `c` going from right to left.
+    ///
+    /// That is, the result is of the form: `f(x1, ...f(xn-1, f(xn, s))...)`.
+    ///
+    pub def foldRight(f: (a, b) -> b & ef, acc: b, c: Nec[a]): b & ef = match viewRight(c) {
+        case OneRight(x)      => f(x, acc)
+        case SomeRight(rs, x) => {
+            let b = f(x, acc);
+            foldRight(f, b, rs)
+        }
+    }
+
+    ///
+    /// Lazily applies `f` to a start value `s` and all elements in `c` going from right to left.
+    /// The function f must be pure.
+    ///
+    /// That is, the result is of the form: `f(x1, ...lazy f(xn-1, lazy f(xn, s))...)`.
+    /// The foldRightLazy function exists to allow early termination of a fold.
+    ///
+    pub def foldRightLazy(f: (a, Lazy[b]) -> b, s: b, c: Nec[a]): b =
+        def loop(l2) = match viewLeft(l2) {
+            case OneLeft(x)      => f(x, lazy s)
+            case SomeLeft(y, ys) => f(y, lazy loop(ys))
+        };
+        loop(c)
+
+    ///
+    /// Collects the results of applying the partial function `f` to every element in `c`.
+    ///
+    pub def filterMap(f: a -> Option[b] & ef, c: Nec[a]): List[b] & ef =
+        def loop(l, k) = match viewLeft(l) {
+            case OneLeft(x)      => match (f(x)) {
+                case Some(a) => k(a :: Nil)
+                case None    => k(Nil)
+            }
+            case SomeLeft(x, rs) => match f(x) {
+                case Some(a) => loop(rs, ks -> k(a :: ks))
+                case None    => loop(rs, k)
+            }
+        };
+        loop(c, ks -> ks as & ef)
+
+    ///
+    /// Returns the first non-None result of applying the partial function `f` to each element of `c`.
+    ///
+    /// Returns `None` if every element of `c` is `None`.
```

Comment:
"if `f(c)` for every element of `c` is `None`

Reply:
Fixed